File: ros2-basic/service_simple/src/node_helloworld_cpp/src/node_server_by_helloworld_interface.cpp
#include "rclcpp/rclcpp.hpp"
#include "node_helloworld_interface/srv/answer.hpp"

class NodeHelloworld : public rclcpp::Node
{
public:
    explicit NodeHelloworld(const std::string &name) : Node(name), count_(0)
    {
        server_ = this->create_service<node_helloworld_interface::srv::Answer>(
            "helloworld_cpp_by_helloworld_interface",
            std::bind(&NodeHelloworld::answer_callback, this,
                      std::placeholders::_1, std::placeholders::_2));
    }

private:
    void answer_callback(
        const node_helloworld_interface::srv::Answer::Request::SharedPtr request,
        node_helloworld_interface::srv::Answer::Response::SharedPtr response)
    {
        auto name = request->name;
        response->message = "[" + name + "] "
                                         "Hello World C++(node_helloworld_interface) Server: " +
                            std::to_string(++count_);
        RCLCPP_INFO(this->get_logger(),
                    "Received request from [%s], response count: %d",
                    name.c_str(), count_);
    }

private:
    int count_;
    rclcpp::Service<node_helloworld_interface::srv::Answer>::SharedPtr server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NodeHelloworld>(
        "node_server_helloworld_cpp_by_helloworld_interface");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2-basic/service_simple/src/node_helloworld_py/node_helloworld_py/node_server_by_helloworld_interface.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from node_helloworld_interface.srv import Answer


class NodeHelloworld(Node):
    def __init__(self, name: str):
        super().__init__(name)
        self.count_ = 0
        # 创建服务
        self.server = self.create_service(
            Answer,
            "helloworld_py_by_helloworld_interface",  # 服务名称标识为 Python
            self.answer_callback,
        )

    def answer_callback(self, request, response):
        name = request.name
        self.count_ += 1
        response.message = f"[{name}] Hello World Python(node_helloworld_interface) Server: {self.count_}"
        self.get_logger().info(
            f"Received request from [{name}], response count: {self.count_}"
        )
        return response


def main(args=None):
    rclpy.init(args=args)
    node = NodeHelloworld("node_server_helloworld_py_by_helloworld_interface")
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()
